docs/2020-05-22-Merging-Rules.md

```diff
@@ -253,12 +253,128 @@ They could be merged into the following axiom:
        ∨ (c = 2 ∧ <k> X:Int /Int Y:Int ~> K </k>)
        ∨ (c = 3 ∧ <k> error("Division by 0") ~> K </k>)
 
+On merging all-path / one-path claims
+-------------------------------------
 
-On merging function-like patterns
----------------------------------
+Since all-path claims are of the form `∀xᵢ.φᵢ → [w] ∃zᵢ.ψᵢ` where
+`[w]` denotes the weak always finally operator, the reasoning above works,
+yielding the following merged rule:
+
+
+    axiom ∀ x1 x2 ... xn c .
+      (c = 1 ∧ l1) ∨ ... (c = n ∧ ln) → [w] (c = 1 ∧ 11) ∨ ... (c = n ∧ rn)
+
+We can then continue merging the lhs part as pointed above.
+
+A similar argument applies for one-path claims.
+
+
+All path reachability proofs
+----------------------------
+(See [All Path Reachability Proofs](2019-03-28-All-Path-Reachability-Proofs))
+
+Assumption: in a goal to be proven, `∀x.φ → [w]∃z.ψ`, `φ` will be assumed to be
+function-like.
+This allows transforming any conjunction `φ ∧ γ` into the equivalent
+`φ ∧ ⌈φ ∧ γ⌉`.
+
+
+
+### Reinterpreting unification
+
+The main idea is to transform `⌈φ ∧ γ⌉` through equivalences to a disjunction
+of conjunctions of equalities between variables and terms representing a
+substitution.  Using equivlaneces guarantees that a MGU is achieved.
+
+In fact, the existing cases for unification stay the same; however, we now need
+to take into account disjunctions and rule choice predicates.
+
+For strongly deterministic definitions (using non-overlapping rules),
+the unification itself will be deterministic, yielding substitutions
+guarded by the choice predicate corresponding to the (only) rule unifying
+with the current configuration.
+
+
+### Reinterpreting the `derivePar` algorithm
+
+
+
+
+Upone merging the one step axioms and the all path claims we interpret the
+all-path algorithm as follows:
+
+__Input__: goal `∀x.φ → [w]∃z.ψ`  and set of tuples `{ (xᵢ,φᵢ,zᵢ,ψᵢ) : 1 ≤ i ≤ n }`
+representing either
+
++ a merged claim `∀x₁ x₂ ... xₙ c . l → [w] r`
++ a merged axiom `∀x₁ x₂ ... xₙ c . l → • ∃ z₁ ... zₙ . r`
+
+where
+
++ `l ∧ (c = i) = φᵢ ∧ c = i` for `1 ≤ i ≤ n`,  and `l ∧ (c = i) = ⊥` if `¬(i = 1) ∧ ... ∧ ¬(i = n)`
++ `r ∧ (c = i) = ψᵢ ∧ c = i` for `1 ≤ i ≤ n`,  and `r ∧ (c = i) = ⊥` if `¬(i = 1) ∧ ... ∧ ¬(i = n)`
+
+__Output:__ `(Goal, goalᵣₑₘ)`
+
+* Let `goalᵣₑₘ := ∀x.(φ ∧ ¬∃x₁ x₂ ... xₙ c .⌈φ ∧ l⌉) → [w]∃z.ψ`
+* Let `Goal := ∀x z₁ z₂ ... zₙ.(∃x₁ x₂ ... xₙ c.r ∧ ⌈φ∧l⌉) → [w]∃z.ψ`
+
+#### Argument for equivalence to the original algorithm
+
+
+Say we want to prove `∀x.φ → [w]∃z.ψ`, and say we want to apply
+circularities, i.e., the merged claim `∀x₁ x₂ ... xₙ c . l → [w] r`.
+
+
+
+
+##### Unification predicate `⌈φ ∧ l⌉`
```

Comment:
This section and the section below are very hard to read in the rendered format because there are no line breaks between equations.